kie-server-parent/kie-server-controller/kie-server-controller-websocket-client/src/main/java/org/kie/server/controller/websocket/client/WebsocketKieServerControllerImpl.java

```diff
@@ -0,0 +1,175 @@
+    
+                        
+                        final KieServerSetup kieServerSetup = new KieServerSetup();
+                        try {
+                            this.client.init(connectAndSyncUrl, config);
+                            CountDownLatch waitLatch = new CountDownLatch(1);
+                            
+                            client.sendWithHandler(serialize(serverInfo), new KieServerSetupMessageHandler(context, waitLatch, kieServerSetup));
+    
+                            boolean received = waitLatch.await(10, TimeUnit.SECONDS);
+                            if (received && kieServerSetup.getContainers() != null) {
+                                // once there is non null list let's return it
+                                return kieServerSetup;
+    
+                            }
+    
+                            break;
+                        } catch (Exception e) {
+                            // let's check all other controllers in case of running in cluster of controllers
+                            logger.warn("Exception encountered while syncing with controller at {} error {}", connectAndSyncUrl, e.getCause() == null ? e.getMessage() : e.getCause().getMessage());
+                            logger.debug("Exception encountered while syncing with controller at {} error {}", connectAndSyncUrl, e.getMessage(), e);
+                        }
+                    } else {
+                        logger.info("Kie Server points to non websocket controller '{}', using default REST mechanism", controllerUrl);
+                        
+                        KieServerSetup kieServerSetup = restController.connectToSingleController(serverInfo, config, controllerUrl);
+                        if (kieServerSetup != null) {
+                            return kieServerSetup;
+                        }
+                    }
+
+                }
+            }
+
+            throw new KieControllerNotConnectedException("Unable to connect to any controller");
+        } else {
+            throw new KieControllerNotDefinedException("Unable to connect to any controller");
+        }
+    }
+
+    @Override
+    public void disconnect(KieServerInfo serverInfo) {
+        if (this.client.isActive()) {
+            this.client.close();
+        } else {            
+            
+            KieServerState currentState = context.getStateRepository().load(KieServerEnvironment.getServerId());
+            Set<String> controllers = currentState.getControllers();
+            KieServerConfig config = currentState.getConfiguration();
+            if (controllers != null && !controllers.isEmpty()) {
+                for (String controllerUrl : controllers) {
+
+                    if (controllerUrl != null && !controllerUrl.isEmpty() && !controllerUrl.toLowerCase().startsWith("ws") || !controllerUrl.toLowerCase().startsWith("wss")) {
```

Comment:
This condition seems bugg:
Since && binds stronger than || the structure is really
`if ((_ && _ && _) || !controllerUrl.toLowerCase().startsWith("wss"))`

which will throw NPE when controllerURL is null. Did you mean to connect this with && or put parenthesis around the two `startsWith` pieces?